### PR TITLE
query_tabular tools for manipulation of tabular files

### PR DESCRIPTION
**Pull request by jj-umn**
Makes use of SQLite database to perform operations on tabular files.
Includes 3 tools: “Query Tabular”, “SQLite to Tabular”, and “Filter
Tabular” with the latter 2 providing a subset of the operations in
Query Tabular.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -2,4 +2,4 @@
 ignore = D,E201,E202,E501
 # For flake8-import-order
 # https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_smarkets.py
-import-order-style = smarkets
+# import-order-style = smarkets
```
Comment (nsoranzo) [inline setup.cfg]: That' what is causing the Travis errors for other tools, can you revert it?

**Comment by jj-umn** [comment]
@nsoranzo   Thanks.  I was trying to get flake8 to run on my Mac and it was complaining about unknown option 'import-order-style'

**Comment by nsoranzo** [comment]
You need to install https://pypi.python.org/pypi/flake8-import-order

File: tools/query_tabular/filter_tabular.xml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    <macros>
+         <import>macros.xml</import>
+    </macros>
+
+    <requirements>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
+    <command><![CDATA[
```
Comment (bgruening) [inline tools/query_tabular/filter_tabular.xml]: `<command detect_errors="exit_code">` and remove the stdio

File: tools/query_tabular/filter_tabular.xml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        python $__tool_directory__/filter_tabular.py
```
Comment (bgruening) [inline tools/query_tabular/filter_tabular.xml]: quotes for the python file as well, the tool_directory could container whitespaces

File: tools/query_tabular/filters.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        for linenum, fields in enumerate(tr):
+            data_lines += 1
+            try:
+                output.write('%s\n' % '\t'.join(fields))
+            except Exception as e:
+                print('Failed at line: %d err: %s' % (linenum, e),
+                      file=sys.stderr)
+    except Exception as e:
+        print('Failed: %s' % (e), file=sys.stderr)
+        exit(1)
```
Comment (bgruening) [inline tools/query_tabular/filters.py]: You could use `exit('Failed: %s' % (e), file=sys.stderr)`

File: tools/query_tabular/sqlite_to_tabular.xml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    <macros>
+         <import>macros.xml</import>
+    </macros>
+
+    <stdio>
+        <exit_code range="1:" level="fatal" description="Error" />
+    </stdio>
+    <command><![CDATA[
+    python $__tool_directory__/sqlite_to_tabular.py
```
Comment (bgruening) [inline tools/query_tabular/sqlite_to_tabular.xml]: quotes

File: tools/query_tabular/sqlite_to_tabular.xml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    <macros>
+         <import>macros.xml</import>
+    </macros>
+
+    <stdio>
+        <exit_code range="1:" level="fatal" description="Error" />
+    </stdio>
+    <command><![CDATA[
+    python $__tool_directory__/sqlite_to_tabular.py 
+    --sqlitedb="$sqlitedb"
```
Comment (bgruening) [inline tools/query_tabular/sqlite_to_tabular.xml]: single quotes

File: tools/query_tabular/filters.py (hunk 4 above)
Comment (nsoranzo) [inline tools/query_tabular/filters.py]: `exit('Failed: %s' % (e))`

**Comment by jj-umn** [comment]
@bgruening @nsoranzo   Thanks for the careful review.  Let me know of any other coding changes I should make.  Is there an equivalent expression in your native languages for "teaching old dogs new tricks?"

**Comment by jj-umn** [comment]
@bgruening @nsoranzo   I think I have this finished.  I tried it with python3 as well as python2.7.  
Could this be squashed on merge?

**Comment by bgruening** [comment]
@jj-umn looks great, lets get this in.

**Comment by nekrut** [comment]
Ready to merge? I want this on main!

**Comment by bgruening** [comment]
@jj-umn can you change the .shed.yaml to be a suite? This would make sense isn't it?

File: tools/query_tabular/query_db.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    exit(0)
+
+
+def run_query(conn, query, outputFile, no_header=False):
+    cur = conn.cursor()
+    results = cur.execute(query)
+    if not no_header:
+        outputFile.write("#%s\n" % '\t'.join(
+            [str(col[0]) for col in cur.description]))
+        # yield [col[0] for col in cur.description]
```
Comment (nsoranzo) [inline tools/query_tabular/query_db.py]: Commented code can be removed maybe?

File: tools/query_tabular/query_tabular.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                     comment_char=comment_char,
+                     load_named_columns=load_named_columns,
+                     filters=filters,
+                     unique_indexes=unique_indexes,
+                     indexes=indexes)
+
+    if options.jsonfile:
+        try:
+            fh = open(options.jsonfile)
+            tdef = json.load(fh)
```
Comment (nsoranzo) [inline tools/query_tabular/query_tabular.py]: Also here you can use the `with` statement.

File: tools/query_tabular/query_tabular.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            if 'tables' in tdef:
+                for ti, table in enumerate(tdef['tables']):
+                    _create_table(ti, table)
+        except Exception as e:
+            exit('Error: %s' % (e))
+
+    query = None
+    if (options.query_file is not None):
```
Comment (nsoranzo) [inline tools/query_tabular/query_tabular.py]: Parentheses are not needed (also below).

File: tools/query_tabular/sqlite_to_tabular.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            outputFile = open(outputPath, 'w')
+        except Exception as e:
+            exit('Error: %s' % (e))
+    else:
+        outputFile = sys.stdout
+
+    query = None
+    if (options.query_file is not None):
+        with open(options.query_file, 'r') as fh:
+            query = ''
```
Comment (nsoranzo) [inline tools/query_tabular/sqlite_to_tabular.py]: I think you can use just:
```
query = fh.read()
```

**Comment by jj-umn** [comment]
@bgruening   Next change will be to make this a suite

**Comment by jj-umn** [comment]
@bgruening If the .shed.yml definition looks correct, I think this should be ready.

**Comment by bgruening** [comment]
Thanks a lot @jj-umn and @nsoranzo!

**Pull request merged**